python/ffsim/qiskit/sim.py

```diff
@@ -767,6 +801,44 @@ def _extract_x_gates(circuit: QuantumCircuit) -> tuple[list[int], QuantumCircuit
     return indices, remaining_circuit
 
 
+def _apply_permutation_gate_vec(
+    vec: np.ndarray,
+    qubit_indices: Sequence[int],
+    perm: Sequence[int],
+    *,
+    norb: int,
+    nelec: int | tuple[int, int],
+    dests: Sequence[int] | None = None,
+) -> np.ndarray:
+    qs = list(qubit_indices)
+    if dests is None:
+        perm = list(perm)
+        inverse_perm = [0] * len(perm)
+        for out_idx, src_idx in enumerate(perm):
+            inverse_perm[src_idx] = out_idx
+        dests = [qs[inverse_perm[i]] for i in range(len(qs))]
+    else:
+        dests = list(dests)
+
+    dest_mask = 0
+    for dest in dests:
+        dest_mask |= 1 << dest
+
+    strings = np.asarray(
+        states.addresses_to_strings(
+            range(len(vec)), norb=norb, nelec=nelec, bitstring_type=BitstringType.INT
+        ),
+        dtype=object,
+    )
+    permuted = strings & ~dest_mask
+    for src, dest in zip(qs, dests):
+        permuted |= ((strings >> src) & 1) << dest
+    indices = states.strings_to_addresses(permuted, norb=norb, nelec=nelec)
+    permuted_vec = np.empty_like(vec)
+    permuted_vec[indices] = vec
+    return permuted_vec
```

Comment:
The bit manipulation tricks in this function can be a bit hard to follow, can you add some comments explaining the logic?